src/backend/distributed/utils/resource_lock.c

```diff
@@ -1060,3 +1062,135 @@ CitusLockTableAclCheck(Oid relationId, LOCKMODE lockmode, Oid userId)
 
 	return aclResult;
 }
+
+
+/*
+ * EnsureCanAcquireLock checks if currect user has the permissions
+ * to acquire a lock on the table and throws an error if the user does
+ * not have the permissions
+ */
+static void
+EnsureCanAcquireLock(Oid relationId, LOCKMODE lockMode)
+{
+	AclResult aclResult = CitusLockTableAclCheck(relationId, lockMode,
+												 GetUserId());
+	if (aclResult != ACLCHECK_OK)
+	{
+		aclcheck_error(aclResult,
+					   get_relkind_objtype(get_rel_relkind(relationId)),
+					   get_rel_name(relationId));
+	}
+}
+
+
+/*
+ * AcquireDistributedLockOnRelations acquire a distributed lock on worker nodes
+ * for given list of relations ids. Worker node list is sorted so that the lock
+ * is acquired in the same order regardless of which node it was run on. Notice
+ * that no lock is acquired on coordinator node if the coordinator is not added
+ * to the metadata.
+ */
+void
+AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
+{
+	const char *lockModeText = LockModeToLockModeText(lockMode);
+
+	UseCoordinatedTransaction();
+
+	StringInfo lockRelationsCommand = makeStringInfo();
+	appendStringInfo(lockRelationsCommand, "%s;\n", DISABLE_DDL_PROPAGATION);
+
+	/*
+	 * In the following loop, when there are foreign tables, we need to switch from
+	 * LOCK * statement to lock_relation_if_exists() and vice versa since pg
+	 * does not support using LOCK on foreign tables.
+	 */
+	bool startedLockCommand = false;
+
+	int lockedRelations = 0;
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIdList)
+	{
+		EnsureCanAcquireLock(relationId, lockMode);
```

Comment:
maybe add comment:
/* we want to prevent under privileged users to trigger establishing connections, that's why we have this additional check. Otherwise, we'd get the errors as soon as we execute the command on any node  */